public/scripts/models/image-upload.js

```diff
@@ -45,23 +39,75 @@ class ImageUpload {
     });
     this.file = file;
     this.maxFileSizeByte = maxFileSizeByte;
-    if (!this.supportedFileTypes.includes(file.type)) {
+  }
+
+  async read() {
+    this.readState = ReadState.Reading;
+    if (!(await this._verifyFileType())) {
       this.readState = ReadState.ErrUnsupportedFileType;
       return;
     }
-    if (this.maxFileSizeByte != null && this.file.size > this.maxFileSizeByte) {
+    if (!this._verifyFileSize()) {
       this.readState = ReadState.ErrExceedMaxFileSize;
       return;
     }
-    this.readState = ReadState.Reading;
-    this.loadDimensionPromise = this._loadDimension();
+    const loadDimensionResult = await this._loadDimension();
+    if (loadDimensionResult.type === "failed") {
+      this.readState = loadDimensionResult.reason;
+      return;
+    }
+    this.readState = ReadState.ReadSuccess;
+  }
+
+  // Cache file type from header such that no need to parse again
+  __fileTypeFromHeader = null;
+  async _getFileTypeFromHeader() {
+    if (this.__fileTypeFromHeader != null) {
+      return this.__fileTypeFromHeader;
+    }
+    const arrayBuffer = await this.file.arrayBuffer();
+    const headerFileTypeBuff = new Uint8Array(arrayBuffer).subarray(0, 4);
+    let fileTypeFromHeader = "";
+    for (let i = 0; i < headerFileTypeBuff.length; i++) {
+      fileTypeFromHeader += String.fromCharCode(
+        parseInt(headerFileTypeBuff[i].toString(16), 16),
+      );
+    }
+    this.__fileTypeFromHeader = fileTypeFromHeader;
+    return this.__fileTypeFromHeader;
+  }
+
+  async _isPsd() {
+    const mimeTypes = ["application/x-photoshop", "image/vnd.adobe.photoshop"];
+    const fileTypesFromHeader = [
+      "8BPS", // https://www.fileformat.info/format/psd/egff.htm
+    ];
+    return (
+      mimeTypes.includes(this.file.type) ||
```

Comment:
So `mimeTypes.includes(this.file.type)` will be true for Non-windows OS & false for window OS?

Reply:
Normally the browser should respect the extension to determine the file type. Windows has some special config. (See notes in https://developer.mozilla.org/en-US/docs/Web/API/Blob/type). So in ideal world we should read header  instead of this field to determine the file type :sosad:

Reply:
> Based on the current implementation, browsers won't actually read the bytestream of a file to determine its media type. It is assumed based on the file extension; a PNG image file renamed to .txt would give "text/plain" and not "image/png". Moreover, blob.type is generally reliable only for common file types like images, HTML documents, audio and video. Uncommon file extensions would return an empty string. Client configuration (for instance, the Windows Registry) may result in unexpected values even for common types. Developers are advised not to rely on this property as a sole validation scheme.

😞